From c23d711239115311eddd40375eeee3d6ca6d3740 Mon Sep 17 00:00:00 2001
From: Patrice Chotard <patrice.chotard@foss.st.com>
Date: Wed, 17 Jan 2024 16:45:59 +0100
Subject: [PATCH] mtd: hyperbus: hyperbus-stm32: Add SFDP calibration mode

Add SFDP calibration mode for JEDEC flashes.

Change-Id: I88e023edc6779d835c5d03fd8c766d68e5783ba0
Signed-off-by: Patrice Chotard <patrice.chotard@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/369131
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Christophe KERELLO <christophe.kerello@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
---
 drivers/mtd/hyperbus/hyperbus-stm32.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

--- a/drivers/mtd/hyperbus/hyperbus-stm32.c
+++ b/drivers/mtd/hyperbus/hyperbus-stm32.c
@@ -5,6 +5,7 @@
  */
 
 #include <linux/mtd/cfi.h>
+#include <linux/mtd/sfdp.h>
 #include <memory/stm32-omi.h>
 
 #define STM32_AUTOSUSPEND_DELAY -1
@@ -120,8 +121,13 @@ static int stm32_hyperbus_check_transfer
 	cfi.interleave = 1;
 	cfi.device_type = CFI_DEVICETYPE_X16;
 
-	ret = cfi_qry_mode_on(0, map, &cfi);
-	cfi_qry_mode_off(0, map, &cfi);
+	if (omi->jedec_flash) {
+		ret = hyperbus_sfdp_mode_on(0, map, &cfi);
+		hyperbus_sfdp_mode_off(0, map, &cfi);
+	} else {
+		ret = cfi_qry_mode_on(0, map, &cfi);
+		cfi_qry_mode_off(0, map, &cfi);
+	}
 
 	return !ret;
 }
